### PR TITLE
BL-4202 Don't ever embed bold or italic fonts

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1513)

### CODE REVIEW

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdDJpEmPqRZlMN4Xy2w:b-ja1hfj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
> ```Smalltalk
> 		{
> 			var group = GetGroupForFont(fontName);
> 			if (group != null && !string.IsNullOrEmpty(group.Normal))
> ```

I notice that you dropped provision for a null group? Probably you have thought about not needing to handle that for some reason, I'm just mentioning it in case.

---

*[src/BloomTests/Publish/ExportEpubTests.cs, line 190 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJQO4J4nISJFBGSYN:-KdDJQO4J4nISJFBGSYO:bb6u11b) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomTests/Publish/ExportEpubTests.cs#L190)):*
> ```Smalltalk
> 
> 			assertThatManifest.HasAtLeastOneMatchForXpath("package/manifest/item[@id='AndikaNewBasic-R' and @href='AndikaNewBasic-R.ttf' and @media-type='application^slash^vnd.ms-opentype']");
> 			assertThatManifest.HasNoMatchForXpath("package/manifest/item[@id='AndikaNewBasic-B' and @href='AndikaNewBasic-B.ttf' and @media-type='application^slash^vnd.ms-opentype']");
> ```

If you do submit a new PR, this new requirement of dropping fonts might be very surprising to a dev in the future, if/when it fails. They'd figure it out eventually. Could use a comment "We drop specialized font faces to save size".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdOqfGBiMKZ19SJzs1L:b-l9cym7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I notice that you dropped provision for a null group? Probably you have thought about not needing to handle that for some reason, I'm just mentioning it in case.
</blockquote></details>

Not following. The change is from returning an (enumerable) group (or an empty array), to YIELD returning one item or (yield) returning nothing at all. Unless I'm misunderstanding how yield works, this will produce zero enumeration results if group is null or has no Normal (and it will have a Normal if we found ANY font with the name), and one if there is a group and a normal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomTests/Publish/ExportEpubTests.cs, line 190 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJQO4J4nISJFBGSYN:-KdR1ZqSeFn8Jwr8hU-e:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomTests/Publish/ExportEpubTests.cs#L190)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

If you do submit a new PR, this new requirement of dropping fonts might be very surprising to a dev in the future, if/when it fails. They'd figure it out eventually. Could use a comment "We drop specialized font faces to save size".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdR7UijV7UadM-BOGzR:bv9wt2m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
> Not following. 

See how the code used to handle the if (group == null) condition? It doesn't anymore.  If group were null, previously, it would return this (odd) new string[0]. Now, instead, it would just not yield anything.

I'm not saying it should be, I'm saying this is a change in behavior that is seemingly unrelated to the change you set out to make, so I'm just pointing it out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by JohnThomson** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdR8fHJYxWb4nfz_Qne:b-yqtu6e) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

> Not following. 

See how the code used to handle the if (group == null) condition? It doesn't anymore.  If group were null, previously, it would return this (odd) new string[0]. Now, instead, it would just not yield anything.

I'm not saying it should be, I'm saying this is a change in behavior that is seemingly unrelated to the change you set out to make, so I'm just pointing it out.
</blockquote></details>

Before, it was not using yield. Therefore, to get no results when group is null, it had to return some other enumerable with no items. Hence the array. Now, using yield return, we get the desired no results by not yielding anything. The only change in behavior is when group is not null: instead of returning the group (typically an enumerable with 4 fonts), we just extract the one thing we want from it and yield that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by andrew-polk** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdRC-BV-Ii002B0xb2L:bawyvvi) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Before, it was not using yield. Therefore, to get no results when group is null, it had to return some other enumerable with no items. Hence the array. Now, using yield return, we get the desired no results by not yielding anything. The only change in behavior is when group is not null: instead of returning the group (typically an enumerable with 4 fonts), we just extract the one thing we want from it and yield that.
</blockquote></details>

In other words, in both the old and new code, a null group resulted in an empty IEnumerable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*

**Comment by hatton** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/FontFileFinder.cs, line 45 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513#-KdDJpEl80m7N9Hy7nj-:-KdRg1y3xblj9XC1KdKg:bj6mwc2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1c3def4fd7207874af92ad9ac96d2da089ce157c/src/BloomExe/Publish/FontFileFinder.cs#L45)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

In other words, in both the old and new code, a null group resulted in an empty IEnumerable.
</blockquote></details>

Oh, weird that it was IEnumberable but not using yield.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1513)*
